File: apps/ios/Pebbles/Features/Glyph/Views/GlyphView.swift
import SwiftUI

/// Canonical glyph chrome component. Renders either the user's glyph strokes
/// or an SF Symbol overlay (scribble for `.carve`, plus for `.create`). Only
/// the carving/empty states (`.carve`, `.create`) draw a dashed 2XL-radius
/// (34pt) frame; glyph-bearing states render chromeless (issue #515).
///
/// Named `GlyphView` (not `Glyph`) because the model type at
/// `Features/Glyph/Models/Glyph.swift` already owns the `Glyph` symbol
/// in this module.
///
/// Visual specification table is in
/// `docs/superpowers/specs/2026-05-17-issue-459-glyph-souls-consistency-design.md` §2.
struct GlyphView: View {
    enum Case {
        case profile      // no frame; glyph in accent.primary
        case carve        // dashed 2pt system.muted; sf.scribble in system.secondary
        case create       // dashed 2pt system.muted; sf.plus in system.muted
        case selected     // no frame; glyph in accent.primary
        case unselected   // no frame; glyph in system.muted
        case `default`    // no frame; glyph in system.secondary
    }

    let `case`: Case
    let strokes: [GlyphStroke]?
    var side: CGFloat = 96

    init(case: Case, strokes: [GlyphStroke]? = nil, side: CGFloat = 96) {
        self.case = `case`
        self.strokes = strokes
        self.side = side
    }

    var body: some View {
        ZStack {
            border
            content
        }
        .frame(width: side, height: side)
    }

    /// Frame is kept only on the carving module and the empty/placeholder
    /// states (`.carve`, `.create`). Glyph-bearing states render chromeless so
    /// the drawing "lives" unframed (issue #515); selection in the picker is
    /// carried by glyph color + the a11y `isSelected` trait instead.
    @ViewBuilder
    private var border: some View {
        let shape = RoundedRectangle(cornerRadius: Spacing.xxl, style: .continuous)
        switch `case` {
        case .carve, .create:
            shape.strokeBorder(
                Color.system.muted,
                style: StrokeStyle(lineWidth: 2, dash: [10, 10])
            )
        case .profile, .selected, .unselected, .default:
            EmptyView()
        }
    }

    @ViewBuilder
    private var content: some View {
        switch `case` {
        case .carve:
            Image(systemName: "scribble")
                .font(.system(size: max(side * 0.4, 18), weight: .regular))
                .foregroundStyle(Color.system.secondary)
        case .create:
            Image(systemName: "plus")
                .font(.system(size: max(side * 0.4, 18), weight: .regular))
                .foregroundStyle(Color.system.muted)
        case .profile, .selected:
            GlyphThumbnail(strokes: strokes ?? [], side: side, strokeColor: Color.accent.primary)
        case .unselected:
            GlyphThumbnail(strokes: strokes ?? [], side: side, strokeColor: Color.system.muted)
        case .default:
            GlyphThumbnail(strokes: strokes ?? [], side: side, strokeColor: Color.system.secondary)
        }
    }
}

#Preview("All cases — light") {
    let strokes = [
        GlyphStroke(d: "M40,40 C80,20 120,20 160,40 S180,120 160,160 S80,180 40,160 S20,80 40,40", width: 6)
    ]
    return ScrollView {
        VStack(spacing: Spacing.lg) {
            HStack(spacing: Spacing.lg) {
                VStack { GlyphView(case: .profile,    strokes: strokes); Text(".profile") }
                VStack { GlyphView(case: .carve);                       Text(".carve") }
                VStack { GlyphView(case: .create);                      Text(".create") }
            }
            HStack(spacing: Spacing.lg) {
                VStack { GlyphView(case: .selected,   strokes: strokes); Text(".selected") }
                VStack { GlyphView(case: .unselected, strokes: strokes); Text(".unselected") }
                VStack { GlyphView(case: .default,    strokes: strokes); Text(".default") }
            }
        }
        .padding(Spacing.lg)
    }
}

#Preview("All cases — dark") {
    let strokes = [
        GlyphStroke(d: "M40,40 C80,20 120,20 160,40 S180,120 160,160 S80,180 40,160 S20,80 40,40", width: 6)
    ]
    return ScrollView {
        VStack(spacing: Spacing.lg) {
            HStack(spacing: Spacing.lg) {
                GlyphView(case: .profile,    strokes: strokes)
                GlyphView(case: .carve)
                GlyphView(case: .create)
            }
            HStack(spacing: Spacing.lg) {
                GlyphView(case: .selected,   strokes: strokes)
                GlyphView(case: .unselected, strokes: strokes)
                GlyphView(case: .default,    strokes: strokes)
            }
        }
        .padding(Spacing.lg)
    }
    .preferredColorScheme(.dark)
}

File: apps/android/app/src/main/kotlin/app/pbbls/android/features/glyph/views/GlyphView.kt
package app.pbbls.android.features.glyph.views

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Icon
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import app.pbbls.android.R
import app.pbbls.android.features.glyph.models.GlyphStroke
import app.pbbls.android.features.path.render.GlyphImage
import app.pbbls.android.theme.PebblesTheme

/**
 * Which chrome + tint a glyph slot renders — ports iOS `GlyphView.Case`
 * (visual spec table: `docs/superpowers/specs/2026-05-17-issue-459-glyph-souls-consistency-design.md` §2).
 * Only the carving/empty states ([CARVE], [CREATE]) draw the dashed
 * `Spacing.xxl`-radius frame; glyph-bearing states render chromeless
 * (issue #515) — selection is carried by glyph color, not a frame.
 */
enum class GlyphViewCase {
    PROFILE, // no frame; glyph in accent.primary
    CARVE, // dashed 2dp system.muted; scribble in system.secondary
    CREATE, // dashed 2dp system.muted; plus in system.muted
    SELECTED, // no frame; glyph in accent.primary
    UNSELECTED, // no frame; glyph in system.muted
    DEFAULT, // no frame; glyph in system.secondary
}

/**
 * Canonical glyph chrome component — ports iOS
 * `Features/Glyph/Views/GlyphView.swift`. Renders either the glyph [strokes]
 * (through the existing [GlyphImage] pipeline) or a placeholder icon overlay
 * (scribble for [GlyphViewCase.CARVE], plus for [GlyphViewCase.CREATE]).
 * Named `GlyphView` for iOS symmetry; the model type owns `Glyph`.
 */
@Composable
fun GlyphView(
    case: GlyphViewCase,
    modifier: Modifier = Modifier,
    strokes: List<GlyphStroke>? = null,
    viewBox: String = "0 0 200 200",
    side: Dp = 96.dp,
) {
    val system = PebblesTheme.colors.system
    val accent = PebblesTheme.colors.accent
    val dashed = case == GlyphViewCase.CARVE || case == GlyphViewCase.CREATE
    val frameColor = system.muted
    val cornerRadius = PebblesTheme.spacing.xxl

    Box(
        modifier =
            modifier
                .size(side)
                .drawBehind {
                    if (dashed) {
                        val strokeWidth = 2.dp.toPx()
                        val dash = 10.dp.toPx()
                        val inset = strokeWidth / 2f
                        drawRoundRect(
                            color = frameColor,
                            topLeft = Offset(inset, inset),
                            size =
                                Size(
                                    size.width - strokeWidth,
                                    size.height - strokeWidth,
                                ),
                            cornerRadius = CornerRadius(cornerRadius.toPx() - inset),
                            style =
                                Stroke(
                                    width = strokeWidth,
                                    pathEffect = PathEffect.dashPathEffect(floatArrayOf(dash, dash)),
                                ),
                        )
                    }
                },
        contentAlignment = Alignment.Center,
    ) {
        when (case) {
            GlyphViewCase.CARVE ->
                PlaceholderIcon(R.drawable.ic_scribble, side, tint = system.secondary)
            GlyphViewCase.CREATE ->
                PlaceholderIcon(R.drawable.ic_plus, side, tint = system.muted)
            GlyphViewCase.PROFILE, GlyphViewCase.SELECTED ->
                GlyphStrokes(strokes, viewBox, accent.primary)
            GlyphViewCase.UNSELECTED ->
                GlyphStrokes(strokes, viewBox, system.muted)
            GlyphViewCase.DEFAULT ->
                GlyphStrokes(strokes, viewBox, system.secondary)
        }
    }
}

@Composable
private fun PlaceholderIcon(
    resId: Int,
    side: Dp,
    tint: Color,
) {
    // iOS sizes the SF Symbol at max(side * 0.4, 18pt).
    val iconSide = if (side * 0.4f > 18.dp) side * 0.4f else 18.dp
    Icon(
        painter = painterResource(resId),
        contentDescription = null,
        tint = tint,
        modifier = Modifier.size(iconSide),
    )
}

@Composable
private fun GlyphStrokes(
    strokes: List<GlyphStroke>?,
    viewBox: String,
    color: Color,
) {
    GlyphImage(
        strokes = strokes.orEmpty(),
        viewBox = viewBox,
        strokeColor = color,
        modifier = Modifier.fillMaxSize(),
    )
}
